.github/workflows/tests.yml

```diff
@@ -0,0 +1,55 @@
+# For more information see: https://help.github.com/actions/language-and-framework-guides/using-python-with-github-actions
+
+name: Test suite
+
+on:
+  pull_request:
+    branches: [ master ]
+
+jobs:
+  lint-and-unit-tests:
+    name: Lint & Unit tests
+    runs-on: ubuntu-latest
+    strategy:
+      fail-fast: true
+      matrix:
+        python-version: [3.6, 3.7, 3.8, 3.9]
+    steps:
+    - uses: actions/checkout@v2
+    - name: Set up Python ${{ matrix.python-version }}
+      uses: actions/setup-python@v2
+      with:
+        python-version: ${{ matrix.python-version }}
+    - name: Install dependencies
+      run: |
+        python -m pip install --upgrade pip
+        python -m pip install tox
+    - name: Run lint checkers
+      run: |
+        tox -e lint
+    - name: Run unit tests
+      run: |
+        tox -e unit
+
+  func-tests:
+    name: Integration tests
+    runs-on: self-hosted
+    needs: lint-and-unit-tests
+    steps:
+      - uses: actions/checkout@v2
+      - name: Set up Python 3.8
+        uses: actions/setup-python@v2
+        with:
+          python-version: 3.8
+      - name: Install dependencies
```

Comment:
I think you should use `actions-operator` to be sure that juju and tox are installed. At the same time it provides the possibility to use GitHub secrets for the credentials and not directly.

Reply:
`actions-operator` is not in the marketplace, and I'm trying to limit which actions can be set up by limiting to those created by GitHub or verified within the marketplace.

OTOH, the secrets I created in the main repository are not available to forked repositories that create a PR against the main repo. This is the reason why I switched to sourcing a file with credentials that are not directly usable for anybody outside of this test env. This together with the limitation that no tests will be run for new contributors until a maintainer approves it should reduce the risks.